File: backend/app/api/routes/datasets.py
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, Depends, UploadFile, File, Form, HTTPException, Query, Request, Body
from sqlalchemy.orm import Session
from pydantic import BaseModel

from backend.app.database.connection import get_db
from backend.app.database.models import User
from backend.app.security.rbac import get_current_user_optional
from backend.app.api.controllers.dataset_controller import DatasetController

router = APIRouter(prefix="/datasets", tags=["datasets"])

class IngestJsonPayload(BaseModel):
    name: str
    sourceType: str = "GENERIC_CSV"
    analysisScope: str = "COMBINED"
    caseId: Optional[str] = None
    rawText: Optional[str] = None
    rows: Optional[List[Dict[str, Any]]] = None

class IngestSamplePayload(BaseModel):
    sampleType: str = "CDR"  # CDR, TRANSACTION, LOCATION, ENTITY
    caseId: Optional[str] = None

@router.get("/summary")
def get_datasets_summary_endpoint(db: Session = Depends(get_db)):
    ctrl = DatasetController(db)
    return ctrl.get_summary()

@router.get("")
def list_datasets_endpoint(
    caseId: Optional[str] = None,
    search: Optional[str] = None,
    sourceType: Optional[str] = Query(None, alias="sourceType"),
    db: Session = Depends(get_db)
):
    ctrl = DatasetController(db)
    return ctrl.list_datasets(case_id=caseId, search=search, source_type=sourceType)

@router.post("/ingest")
async def ingest_dataset_file_endpoint(
    request: Request,
    file: Optional[UploadFile] = File(None),
    name: Optional[str] = Form(None),
    sourceType: Optional[str] = Form("GENERIC_CSV"),
    caseId: Optional[str] = Form(None),
    analysisScope: Optional[str] = Form("COMBINED"),
    user: Optional[User] = Depends(get_current_user_optional),
    db: Session = Depends(get_db)
):
    ctrl = DatasetController(db)
    file_content = None
    file_name = None

    if file:
        file_content = await file.read()
        file_name = file.filename
        if not name:
            name = file.filename.rsplit(".", 1)[0].replace("_", " ").title()

    if not name:
        name = "Forensic Ingestion Dataset"

    client_ip = request.client.host if request and request.client else "127.0.0.1"

    try:
        dataset = ctrl.ingest_dataset(
            name=name,
            source_type=sourceType or "GENERIC_CSV",
            file_content=file_content,
            file_name=file_name,
            case_id=caseId,
            analysis_scope=analysisScope or "COMBINED",
            user_id=user.id if user else None,
            role=user.role if user else "INVESTIGATOR",
            ip=client_ip
        )
        return {"success": True, "dataset": dataset}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/ingest-json")
def ingest_dataset_json_endpoint(
    payload: IngestJsonPayload,
    request: Request,
    user: Optional[User] = Depends(get_current_user_optional),
    db: Session = Depends(get_db)
):
    ctrl = DatasetController(db)
    client_ip = request.client.host if request and request.client else "127.0.0.1"
    try:
        dataset = ctrl.ingest_dataset(
            name=payload.name,
            source_type=payload.sourceType,
            raw_text=payload.rawText,
            raw_rows=payload.rows,
            case_id=payload.caseId,
            analysis_scope=payload.analysisScope,
            user_id=user.id if user else None,
            role=user.role if user else "INVESTIGATOR",
            ip=client_ip
        )
        return {"success": True, "dataset": dataset}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/sample")
def ingest_sample_dataset_endpoint(
    payload: IngestSamplePayload,
    user: Optional[User] = Depends(get_current_user_optional),
    db: Session = Depends(get_db)
):
    ctrl = DatasetController(db)
    try:
        dataset = ctrl.ingest_sample(
            sample_type=payload.sampleType,
            case_id=payload.caseId,
            user_id=user.id if user else None,
            role=user.role if user else "INVESTIGATOR"
        )
        return {"success": True, "dataset": dataset}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{dataset_id}")
def get_dataset_detail_endpoint(dataset_id: str, db: Session = Depends(get_db)):
    ctrl = DatasetController(db)
    ds = ctrl.get_dataset(dataset_id)
    if not ds:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return ds

@router.get("/{dataset_id}/records")
def get_dataset_records_endpoint(
    dataset_id: str,
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
    search: Optional[str] = None,
    db: Session = Depends(get_db)
):
    ctrl = DatasetController(db)
    return ctrl.get_records(dataset_id, limit=limit, offset=offset, search=search)

@router.delete("/{dataset_id}")
def delete_dataset_endpoint(
    dataset_id: str,
    user: Optional[User] = Depends(get_current_user_optional),
    db: Session = Depends(get_db)
):
    ctrl = DatasetController(db)
    success = ctrl.delete_dataset(
        dataset_id,
        user_id=user.id if user else None,
        role=user.role if user else "INVESTIGATOR"
    )
    if not success:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return {"success": True, "message": "Dataset deleted successfully"}

File: frontend/src/services/datasets.ts
import { apiRequest } from './api';

export interface DatasetItem {
  id: string;
  name: string;
  sourceType: 'CSV' | 'CDR' | 'TRANSACTION' | 'LOCATION' | 'GENERIC_CSV' | string;
  fileName?: string;
  status: string;
  recordCount: number;
  analysisScope: string;
  caseId?: string;
  createdAt: string;
  updatedAt?: string;
}

export interface DatasetSummary {
  totalDatasets: number;
  totalRecords: number;
  sourceBreakdown: Record<string, number>;
}

export interface DatasetRecordItem {
  id: string;
  rowIndex: number;
  raw: Record<string, any>;
  normalized: Record<string, any>;
  matchStatus: string;
  matchConfidence: number;
  matchReasons?: string;
  createdAt: string;
}

export interface DatasetRecordsResponse {
  total: number;
  limit: number;
  offset: number;
  records: DatasetRecordItem[];
}

export interface IngestJsonPayload {
  name: string;
  sourceType: string;
  analysisScope?: string;
  caseId?: string;
  rawText?: string;
  rows?: Array<Record<string, any>>;
}

export const datasetService = {
  list: async (caseId?: string, search?: string, sourceType?: string): Promise<DatasetItem[]> => {
    const params = new URLSearchParams();
    if (caseId) params.append('caseId', caseId);
    if (search) params.append('search', search);
    if (sourceType && sourceType !== 'ALL') params.append('sourceType', sourceType);
    const qs = params.toString() ? `?${params.toString()}` : '';
    return apiRequest<DatasetItem[]>(`/datasets${qs}`);
  },

  getSummary: async (): Promise<DatasetSummary> => {
    return apiRequest<DatasetSummary>('/datasets/summary');
  },

  getById: async (datasetId: string): Promise<DatasetItem> => {
    return apiRequest<DatasetItem>(`/datasets/${encodeURIComponent(datasetId)}`);
  },

  getRecords: async (
    datasetId: string,
    limit: number = 50,
    offset: number = 0,
    search?: string
  ): Promise<DatasetRecordsResponse> => {
    const params = new URLSearchParams();
    params.append('limit', limit.toString());
    params.append('offset', offset.toString());
    if (search) params.append('search', search);
    return apiRequest<DatasetRecordsResponse>(`/datasets/${encodeURIComponent(datasetId)}/records?${params.toString()}`);
  },

  ingestFile: async (formData: FormData): Promise<{ success: boolean; dataset: DatasetItem }> => {
    return apiRequest<{ success: boolean; dataset: DatasetItem }>('/datasets/ingest', {
      method: 'POST',
      body: formData,
      // Note: when body is FormData, apiRequest does not set Content-Type so browser sets boundary multipart automatically
    });
  },

  ingestJson: async (payload: IngestJsonPayload): Promise<{ success: boolean; dataset: DatasetItem }> => {
    return apiRequest<{ success: boolean; dataset: DatasetItem }>('/datasets/ingest-json', {
      method: 'POST',
      body: JSON.stringify(payload),
    });
  },

  ingestSample: async (
    sampleType: 'CDR' | 'TRANSACTION' | 'LOCATION' | 'ENTITY',
    caseId?: string
  ): Promise<{ success: boolean; dataset: DatasetItem }> => {
    return apiRequest<{ success: boolean; dataset: DatasetItem }>('/datasets/sample', {
      method: 'POST',
      body: JSON.stringify({ sampleType, caseId }),
    });
  },

  delete: async (datasetId: string): Promise<{ success: boolean; message: string }> => {
    return apiRequest<{ success: boolean; message: string }>(`/datasets/${encodeURIComponent(datasetId)}`, {
      method: 'DELETE',
    });
  },
};
